bin/check

```diff
@@ -19,172 +19,71 @@ set -euo pipefail
-    # `collection.is_empty()`, but the argument that `.is_empty()` is faster is
-    # dubious, and it often breaks symmetry in a chain of length comparisons,
-    # like `if c.len() == 0 { .. } else if c.len() == 1 { .. } else { .. }`.
-    clippy::len_zero
-
-    # Upstream description: https://rust-lang.github.io/rust-clippy/master/#let_and_return
-    #
-    # clippy::let_and_return protects against creating a let binding that is
-    # returned from a block without otherwise being used, like so:
-    #     {
-    #       let x = String::new();
-    #       // TODO(frank): figure out why sorting here is broken.
-    #       // x.sort()
-    #       x
-    #     }
-    #
-    # This situation can arise naturally when commenting out intermediate code,
-    # like in the example above, and working around the lint requires corrupting
-    # the original intent of the code.
-    clippy::let_and_return
-
-    # Upstream description: https://rust-lang.github.io/rust-clippy/master/#match_bool
-    #
-    # clippy::match_bool prohibits matching in bools, like so
-    #
-    #     match cond {
-    #       true => (),
-    #       false => (),
-    #     }
-    #
-    # and instead mandates use of an if/else statement. This is needlessly
-    # nitpicky. For one, matching on bools uses one fewer lines than using an
-    # if/else statement; for another, the author may be maintaining a consistent
-    # style with similar blocks of nearby code.
-    clippy::match_bool
+    # clippy::mutable_key_type disallows using internally mutable types as keys in `HashMap`,
+    # because their order could change. This is a good lint in principle, but its current
+    # implementation is too strict -- it disallows anything containing an `Arc` or `Rc`,
+    # for example.
+    clippy::mutable_key_type
 
-    # Upstream description: https://github.com/rust-lang/rust-clippy/issues/5991
-    #
     # This lint has too many false positives in Rust v1.47.0.
     # See: https://github.com/rust-lang/rust-clippy/issues/5991
     clippy::needless-collect
 
-    # Upstream description: https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
-    #
-    # Lifetime elision sometimes gives surprising results, and especially in
-    # unsafe code it's useful to spell out lifetimes explicitly to be sure.
-    clippy::needless-lifetimes
+    # Unstable sort is not strictly better than sort, notably on partially
+    # sorted inputs.
+    clippy::stable-sort-primitive
+)
 
-    # Upstream description: https://rust-lang.github.io/rust-clippy/master/#needless_range_loop
-    #
-    # clippy::needless_range_loop encourages use of `for item in collection`
-    # instead of `for i in 0..collection.len()`. The lint is overly aggressive,
-    # however, and often forces the creation of complicated stacks of iterators
-    # when the for loop is clearer.
-    clippy::needless_range_loop
+extra_lints=(
+    # All compiler warnings should be considered errors.
```

Comment:
You have a bunch of good ones here already, but we can probably safely add more (which I volunteer to take the time to do, after this lands).

Discussion in slack: https://materializeinc.slack.com/archives/CMH6PG4CW/p1614229204132900